### PR TITLE
Cleaning reducers - [MOD-7251]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

General code cleanup for the `GROUPBY` reducers.

#### Main Changes
- `group_by.c` - since 2.8, we validate correctly that each reducer has a unique destination key, so we don't need to override `writeGroupValues` after every reducer.
- `first_value.c` - use `RS_NullVal` (today's convention) instead of `RS_StaticNull`.
- `minmax.c` - split `min` and `max` callbacks.
- `to_list.c` - replace triemap with a dictionary.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/aggregate/reducers/to_list.c [hunk 1]
```diff
@@ -6,75 +6,70 @@
...
+  .keyDestructor = destructor_RSValue,
+  .valDestructor = NULL,
+};
 
 static void *tolistNewInstance(Reducer *rbase) {
-  tolistCtx *ctx = Reducer_BlkAlloc(rbase, sizeof(*ctx), 100 * sizeof(*ctx));
-  ctx->values = NewTrieMap();
-  ctx->srckey = rbase->srckey;
-  return ctx;
+  dict *values = dictCreate(&RSValueSet, NULL);
```
Comment (DvirDukhan) [inline src/aggregate/reducers/to_list.c:33]: why the change from trie to dict? I think this is the most important question in this PR

File: src/aggregate/reducers/to_list.c [hunk 2]
```diff
@@ -6,75 +6,70 @@
...
-  RSValue *val;
-  RSValue **arr = RSValue_AllocateArray(tlc->values->cardinality);
-  size_t i = 0;
-  while (TrieMapIterator_Next(it, &c, &l, (void **)&val)) {
-    if (val) {
-      arr[i++] = RSValue_IncrRef(val);
-    }
+  dict *values = ctx;
+  size_t len = dictSize(values);
+  dictIterator *it = dictGetIterator(values);
```
Comment (DvirDukhan) [inline src/aggregate/reducers/to_list.c:59]: anything can be affected here? list creation order?

File: src/aggregate/group_by.c [hunk 3]
```diff
@@ -294,12 +279,12 @@ Grouper *Grouper_New(const RLookupKey **srckeys, const RLookupKey **dstkeys, siz
...
 
-  g->srckeys = rm_calloc(nkeys, sizeof(*g->srckeys));
-  g->dstkeys = rm_calloc(nkeys, sizeof(*g->dstkeys));
   g->nkeys = nkeys;
-  for (size_t ii = 0; ii < nkeys; ++ii) {
-    g->srckeys[ii] = srckeys[ii];
-    g->dstkeys[ii] = dstkeys[ii];
+  if (nkeys) {
+    g->srckeys = rm_malloc(nkeys * sizeof(*g->srckeys));
+    g->dstkeys = rm_malloc(nkeys * sizeof(*g->dstkeys));
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/aggregate/group_by.c:285]: `rm_free` handles NULLs. Also `malloc` and `calloc` may return NULL if 0 bytes were requested

File: src/aggregate/reducers/count_distinct.c [hunk 4]
```diff
@@ -164,14 +162,11 @@ Reducer *RDCRHLL_New(const ReducerOptions *options) {
   return newHllCommon(options, 1);
 }
 
-typedef struct {
-  const RLookupKey *srckey;
-  struct HLL hll;
-} hllSumCtx;
+typedef struct HLL hllSumCtx;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/aggregate/reducers/count_distinct.c:165]: Mostly to keep the context semantic that is being used across the reducers, and also the struct is not typedef-ed yet

File: src/aggregate/reducers/first_value.c [hunk 5]
```diff
@@ -83,7 +78,7 @@ static RSValue *fvFinalize(Reducer *parent, void *ctx) {
   if (fvx->value) {
     return RSValue_IncrRef(fvx->value);
   } else {
-    return &RS_StaticNull;
+    return RS_NullVal();
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/aggregate/reducers/first_value.c:81]: I'm not sure if it is reachable or not. Wasn't covered before.
There is no intentional difference (at least not today when `RS_NullVal()` is also static, even though the comment on it says it returns a new null instance). This was almost the only place that used this value and not `RS_NullVal()`. I removed this null variant completely

File: src/value.h [hunk 6]
```diff
@@ -267,22 +265,6 @@ static inline int RSValue_IsNull(const RSValue *value) {
   return 0;
 }
 
-/* Make sure a value can be long lived. If the underlying value is a volatile string that might go
- * away in the next iteration, we copy it at that stage. This doesn't change the ref count.
- * A volatile string usually comes from a block allocator and is not freed in RSVAlue_Free, so just
- * discarding the pointer here is "safe" */
-static inline RSValue *RSValue_MakePersistent(RSValue *v) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/value.h:274]: `RSString_Volatile` is not in use, and this function was only handling this type of string value. It was used in `to_list` but had no effect

File: src/aggregate/reducers/random_sample.c [hunk 7]
```diff
@@ -49,19 +49,12 @@ static int sampleAdd(Reducer *rbase, void *ctx, const RLookupRow *srcrow) {
 
 static RSValue *sampleFinalize(Reducer *rbase, void *ctx) {
   rsmplCtx *sc = ctx;
-  RSMPLReducer *r = (RSMPLReducer *)rbase;
-  size_t len = MIN(r->len, sc->seen);
-  RSValue *ret = sc->samplesArray;
-  sc->samplesArray = NULL;
-  return ret;
+  return RSValue_IncrRef(sc->samplesArray);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/aggregate/reducers/random_sample.c:52]: Utilizing the refcount mechanism instead of giving away the pointer. It just makes the code flow simpler.
The `Finalize` function gives a reference to the caller, and freeing the reducer instance always includes releasing the instance reference. This is the common way of passing ownership over RSValue in other places in the code as well

File: src/aggregate/reducers/random_sample.c [hunk 8]
```diff
@@ -49,19 +49,12 @@ static int sampleAdd(Reducer *rbase, void *ctx, const RLookupRow *srcrow) {
...
 }
 
 static void sampleFreeInstance(Reducer *rbase, void *p) {
   rsmplCtx *sc = p;
-  RSMPLReducer *r = (RSMPLReducer *)rbase;
-  if (sc->samplesArray) {
-    RSValue_Decref(sc->samplesArray);
-  }
+  RSValue_Decref(sc->samplesArray);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/aggregate/reducers/random_sample.c:57]: No. The reducer instance always has its reference to the array, even if `Finalize` was called. So we always have to release it here

File: src/aggregate/reducers/to_list.c (hunk 1 above)
Comment (GuyAv46) [reply src/aggregate/reducers/to_list.c:33]: There was no good reason to abuse the triemap (designed for strings) to store the `uint64_t` hash that was calculated manually and then passed as a string. Also later iterating over the trie with an empty string so all the results will be found. Moving to a dictionary simplifies the flow (calling `dictAdd` instead of calculating the hash, checking if it already exists, and only then inserting the value). The dictionary handles all of the logic of taking references to new values, freeing them in the end, calculating the hash, and so on. Also iterating over it makes more sense.

File: src/aggregate/reducers/to_list.c (hunk 2 above)
Comment (GuyAv46) [reply src/aggregate/reducers/to_list.c:59]: Previously the order was lexicographic over the hash value of the values when viewed as a char[8]. I think the dictionary also iterates over the values according to the hash value, but there was no guarantee for any specific order anyway

File: src/aggregate/reducers/first_value.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/aggregate/reducers/first_value.c:81]: maybe change the comment of `RS_NullVal()`?

File: src/aggregate/reducers/random_sample.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/aggregate/reducers/random_sample.c:52]: Is the a comment documenting the ownership transfer?

**Review by DvirDukhan — APPROVED** [review]
